pil-analyzer/src/condenser.rs

```diff
@@ -810,6 +812,85 @@ fn try_to_function_value_definition<T: FieldElement>(
     }))
 }
 
+fn try_algebraic_expression_to_expression<T: FieldElement>(
+    e: &AlgebraicExpression<T>,
+) -> Result<Expression, EvalError> {
+    Ok(match e {
+        AlgebraicExpression::Reference(AlgebraicReference {
+            name,
+            poly_id: _,
+            next: false,
+        }) => Expression::Reference(
+            SourceRef::unknown(),
+            Reference::Poly(PolynomialReference {
+                name: name.clone(),
+                type_args: None,
+            }),
+        ),
+
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op, right }) => {
+            Expression::BinaryOperation(
+                SourceRef::unknown(),
+                BinaryOperation {
+                    left: Box::new(
+                        try_value_to_expression(&Value::<T>::Expression((**left).clone())).unwrap(),
```

Comment:
```suggestion
                        try_algebraic_expression_to_expression(left)?
```

Reply:
When you need to clone something of unbounded size, there should be a better solution. Not always, but most of the time.

Reply:
Also you could try `BinaryOperation{ ... }.into()` instead of constructing `Expression::BinaryOperation` explicitly.